### PR TITLE
QANER tutorial added

### PR DESCRIPTION
**Pull request by SergeyAksenovSA**

### CODE REVIEW

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 1]
```diff
@@ -0,0 +1,1341 @@
...
+   "id": "eeca537e-ba33-4389-a5ba-bbc97b062515",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "! pip install superannotate==4.4.7 #SA SDK installation\n",
+    "! pip install boto3 # install boto3 client\n",
+    "! pip install transformers # HuggingFace transformers\n",
+    "! pip install seqeval # model evaluation"
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb]: Let's fix versions

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 2]
```diff
@@ -0,0 +1,1341 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for subset_name in subset_names:\n",
+    "    print(f\"Loading {subset_name} docs\")\n",
+    "    save_dir = f'./{subset_name}_sa_docs'\n",
+    "    if not os.path.exists(save_dir):\n",
+    "        os.mkdir(save_dir)\n",
+    "    for file_key in tqdm(data_links_dict[subset_name]):\n",
+    "        if not '.txt' in file_key:\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb]: if not file_key.endswith('.txt'):

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 3]
```diff
@@ -0,0 +1,1341 @@
...
+    "            continue\n",
+    "        filename = os.path.basename(file_key)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:415]: Let's add some comment on it

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 4]
```diff
@@ -0,0 +1,1341 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "58d74089",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "prompt_mapper = {'CASE_NUMBER': 'case number',\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb]: add extra \t under, so it will look more structured

Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb]: prompt_mapper = PROMPT_MAPPER

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 5]
```diff
@@ -0,0 +1,1341 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "789fc352",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "Instance = namedtuple(\"Instance\", [\"context\", \"question\", \"answer\"])\n",
+    "\n",
+    "class Dataset(torch.utils.data.Dataset):\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb]: Dataset -> QADataset

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 6]
```diff
@@ -0,0 +1,1341 @@
...
+    "\n",
+    "    def __getitem__(self, idx: int):\n",
+    "        return self.dataset[idx]\n",
+    "\n",
+    "    def _prepare_dataset(self,\n",
+    "                         qa_sentences,\n",
+    "                         qa_labels):\n",
+    "\n",
+    "        dataset = []\n",
+    "        for sentence, labels in tqdm(zip(qa_sentences, qa_labels),desc=\"prepare_dataset\"):\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb]: Uhhhhm, maybe some comments? What is happening here?))

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 7]
```diff
@@ -0,0 +1,1341 @@
...
+    "            for label_tag, label_name in self.prompt_mapper.items():\n",
+    "                question_prompt = f\"What is the {label_name}?\"\n",
+    "\n",
+    "                answer_list = []\n",
+    "                for span in labels:\n",
+    "                    if span['label'] == label_tag:\n",
+    "                        answer_list.append(span)\n",
+    "\n",
+    "                if len(answer_list) == 0:\n",
+    "                    empty_span = {\"token\" : \"\",\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:561]: What is SPAN? What is Prompt? What is Instance? 
Add more text about terminology and a few words on how it's working

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 8]
```diff
@@ -0,0 +1,1341 @@
...
+    "                            context=sentence,\n",
+    "                            question=question_prompt,\n",
+    "                            answer=answer,\n",
+    "                        )\n",
+    "                        dataset.append(instance)\n",
+    "\n",
+    "        return dataset\n",
+    "\n",
+    "    \n",
+    "class Collator:\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:595]: What is collator? Why do we need it?

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 9]
```diff
@@ -0,0 +1,1341 @@
...
+    "                    res_end_token_pos = i  \n",
+    "                    done = True\n",
+    "                    break\n",
+    "        if special_tokens_cnt > 2:\n",
+    "            res_end_token_pos = len(offset_mapping) - 1  \n",
+    "            res_start_token_pos = 0\n",
+    "            done = True\n",
+    "            \n",
+    "        assert done\n",
+    "        return res_start_token_pos, res_end_token_pos"
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:664]: For each function add comments on what is it and how it's working

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 10]
```diff
@@ -0,0 +1,1341 @@
...
+    "        return dataset\n",
+    "\n",
+    "    \n",
+    "class Collator:\n",
+    "    def __init__(self, tokenizer, tokenizer_kwargs):\n",
+    "        self.tokenizer = tokenizer\n",
+    "        self.tokenizer_kwargs = tokenizer_kwargs\n",
+    "\n",
+    "    def __call__(self, batch):\n",
+    "        context_list = []\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:601]: Comments, comments, comments

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 11]
```diff
@@ -0,0 +1,1341 @@
...
+    "    for span_true, span_pred in zip(spans_true_batch, spans_pred_batch_top_1):\n",
+    "        span_pred = span_pred[0]\n",
+    "        i = entity_mapper[span_true.label]\n",
+    "        j = entity_mapper[span_pred.label]\n",
+    "        confusion_matrix_true_denominator[i] += 1\n",
+    "        confusion_matrix_pred_denominator[j] += 1\n",
+    "        if span_true == span_pred:\n",
+    "            ner_confusion_matrix[i, j] += 1\n",
+    "            \n",
+    "    assert (confusion_matrix_true_denominator.sum() == confusion_matrix_pred_denominator.sum())\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:782]: What is going on?

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 12]
```diff
@@ -0,0 +1,1341 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "48222fba",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_top_valid_spans(context_list,\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:827]: What is this? Why do we need this function? How it''s working?

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb [hunk 13]
```diff
@@ -0,0 +1,1341 @@
...
+    "                                            spans_pred_batch_top_1=spans_pred_batch_top_1,\n",
+    "                                            prompt_mapper=dataloader.dataset.prompt_mapper)\n",
+    "            for metric_name, metric_value in batch_metrics.items():\n",
+    "                batch_metrics_list[metric_name].append(metric_value)\n",
+    "        avg_loss = np.mean(epoch_loss)\n",
+    "        print(f\"Test loss:  {avg_loss}\\n\")\n",
+    "\n",
+    "        for metric_name, metric_value_list in batch_metrics_list.items():\n",
+    "            metric_value = np.mean(metric_value_list)\n",
+    "            print(f\"Test {metric_name}: {metric_value}\\n\")"
```
Comment (konyrevdmitriy) [inline named_entity_recognition/named_entity_recognition_question_answering.ipynb:1014]: Comments, comments, comments

**Review by konyrevdmitriy — COMMENTED** [review]
More comments more explanation, user will not understand what is happening

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb (hunk 2 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition/named_entity_recognition_question_answering.ipynb]: +

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb (hunk 4 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition/named_entity_recognition_question_answering.ipynb]: +

Comment (SergeyAksenovSA) [reply named_entity_recognition/named_entity_recognition_question_answering.ipynb]: +

File: named_entity_recognition/named_entity_recognition_question_answering.ipynb (hunk 5 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition/named_entity_recognition_question_answering.ipynb]: +

**Pull request merged**